src/v/cluster/partition_balancer_backend.cc

```diff
@@ -60,29 +61,89 @@ partition_balancer_backend::partition_balancer_backend(
   , _tick_interval(std::move(tick_interval))
   , _movement_batch_size_bytes(std::move(movement_batch_size_bytes))
   , _max_concurrent_actions(std::move(max_concurrent_actions))
+  , _concurrent_moves_drop_threshold(std::move(moves_drop_threshold))
   , _segment_fallocation_step(std::move(segment_fallocation_step))
   , _timer([this] { tick(); }) {}
 
 void partition_balancer_backend::start() {
-    _timer.arm(_tick_interval());
+    _topic_table_updates = _state.topics().register_lw_notification(
+      [this]() { on_topic_table_update(); });
+    _member_updates = _state.members().register_members_updated_notification(
+      [this](model::node_id n, model::membership_state state) {
+          on_members_update(n, state);
+      });
+    maybe_rearm_timer();
     vlog(clusterlog.info, "partition balancer started");
 }
 
+void partition_balancer_backend::maybe_rearm_timer(bool now) {
+    if (_gate.is_closed()) {
+        return;
+    }
+    auto schedule_at = now ? clock_t::now() : clock_t::now() + _tick_interval();
+    auto duration_ms = [](clock_t::time_point time_point) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+                 time_point - clock_t::now())
+          .count();
+    };
+    if (_timer.armed()) {
+        schedule_at = std::min(schedule_at, _timer.get_timeout());
+        _timer.rearm(schedule_at);
+        vlog(
+          clusterlog.debug,
+          "Tick rescheduled to run in: {}ms",
+          duration_ms(schedule_at));
+    } else if (_lock.waiters() == 0) {
```

Comment:
> we'll have to wait for the next timer-based tick instead.

no, that's wrong - as the number of waiters doesn't include the current mutex holder, we'll rearm the timer, it will immediately fire, the spawned fiber will block on the mutex and after it becomes available, we'll run the balancer again.